File: python-algorithm/algo_06_backtrack/L37-h.py
#--coding:utf-8--
'''
@Time   : 2020/8/25
@Author : Heng Li
@Email  : liheng_lulia@163.com
'''

class Solution:
    def __init__(self, **kwargs):
        self.row = []
        self.col = []
        self.box = []
        self.blank = []
        self.res = []

    # def solveSudoku(self, board: List[List[str]]) -> None:
    def solveSudoku(self, board) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        # 里面做成10是为了映射到1-9具体每个数字
        # 记录每行每个数是否出现过；每列...;每个九宫格...
        self.row = [[0 for j in range(10)] for i in range(9)]
        self.col = [[0 for j in range(10)] for i in range(9)]
        self.box = [[0 for j in range(10)] for i in range(9)]
        for i in range(9):
            for j in range(9):
                if board[i][j]=='.':
                    self.blank.append([i,j])
                else:
                    curNum = int(board[i][j])
                    self.row[i][curNum] = 1
                    self.col[j][curNum] = 1
                    self.box[j//3+(i//3)*3][curNum] = 1
        self.dfs(board, 0, [])
        # 题目说假设只有一种解，即res里面只含有一组解
        for i in range(len(self.res[0])):
            board[self.blank[i][0]][self.blank[i][1]] = str(self.res[0][i])

    def dfs(self, board, idx, cur_ls):
        if idx == len(self.blank):
            self.res.append(cur_ls[:])
            return
        for num in range(1,10):
            t_r = self.blank[idx][0]
            t_c = self.blank[idx][1]
            if not self.row[t_r][num] and not self.col[t_c][num] and \
            not self.box[t_c//3 + (t_r//3)*3][num]:
                cur_ls.append(num)
                self.row[t_r][num] = 1
                self.col[t_c][num] = 1
                self.box[t_c//3 + (t_r//3)*3][num] = 1
                self.dfs(board, idx+1, cur_ls)
                cur_ls.pop()
                self.row[t_r][num] = 0
                self.col[t_c][num] = 0
                self.box[t_c//3 + (t_r//3)*3][num] = 0


if __name__ == "__main__":
    board =[['5','3','.','.','7','.','.','.','.'],
            ['6','.','.','1','9','5','.','.','.'],
            ['.','9','8','.','.','.','.','6','.'],
            ['8','.','.','.','6','.','.','.','3'],
            ['4','.','.','8','.','3','.','.','1'],
            ['7','.','.','.','2','.','.','.','6'],
            ['.','6','.','.','.','.','2','8','.'],
            ['.','.','.','4','1','9','.','.','5'],
            ['.','.','.','.','8','.','.','7','9']]

    Solution().solveSudoku(board)
    for i in board:
        print(i)

File: python-algorithm/algo_06_backtrack/L37-h.cpp

#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    vector<vector<int>> pos;
    vector<vector<char>> res;
    void solveSudoku(vector<vector<char>>& board) {
        //初始化vector放在构造函数或者函数，不能直接在外面初始化
        vector<vector<int>> row(9, vector<int>(10,0));
        vector<vector<int>> col(9, vector<int>(10,0));
        vector<vector<int>> box(9, vector<int>(10,0));
        for(int i=0; i<9; ++i)
            for(int j=0; j<9; ++j){
                if(board[i][j] == '.')
                    pos.push_back(vector<int>{i,j});
                else{
                    int curNumber = board[i][j]-'0';
                    row[i][curNumber] = 1;
                    col[j][curNumber] = 1;
                    box[j/3 + (i/3)*3][curNumber] = 1;
                }
            }
        vector<char> cur;
        dfs(board, 0, cur, row, col, box);
        for(auto v:res){
            for(int i=0; i<v.size(); ++i)
                board[pos[i][0]][pos[i][1]] = v[i] ;
        }

    }

    void dfs(vector<vector<char>>& board, int idx, vector<char>& cur,
             vector<vector<int>>&row,vector<vector<int>>&col,vector<vector<int>>&box){
        if(idx == pos.size()){
            res.push_back(cur);
            return;
        }
        for(int num=1; num<=9; ++num){
            int t_r = pos[idx][0];
            int t_c = pos[idx][1];
            if(!row[t_r][num] && !col[t_c][num] && !box[t_c/3 + (t_r/3)*3][num]){
                char c = to_string(num)[0];
                cur.push_back(c);
                row[t_r][num] = 1;
                col[t_c][num] = 1;
                box[t_c/3 + (t_r/3)*3][num] = 1;
                dfs(board, idx+1, cur, row, col, box);
                cur.pop_back();
                row[t_r][num] = 0;
                col[t_c][num] = 0;
                box[t_c/3 + (t_r/3)*3][num] = 0;
            }
        }
    }
};


int main(){
    vector<vector<char>> b = {{'5','3','.','.','7','.','.','.','.'},
                              {'6','.','.','1','9','5','.','.','.'},
                              {'.','9','8','.','.','.','.','6','.'},
                              {'8','.','.','.','6','.','.','.','3'},
                              {'4','.','.','8','.','3','.','.','1'},
                              {'7','.','.','.','2','.','.','.','6'},
                              {'.','6','.','.','.','.','2','8','.'},
                              {'.','.','.','4','1','9','.','.','5'},
                              {'.','.','.','.','8','.','.','7','9'}};
    Solution().solveSudoku(b);
}